internal/client/client_async.go

```diff
@@ -0,0 +1,196 @@
+		ctx = opentracing.ContextWithSpan(ctx, spanRPC)
+	}
+
+	useCodec := c.option != nil && c.option.codec != nil
+	if useCodec {
+		req, err = c.option.codec.EncodeRequest(req)
+		if err != nil {
+			cb.Invoke(nil, err)
+			return
+		}
+	}
+	tikvrpc.AttachContext(req, req.Context)
+
+	// TODO(zyguan): If the client created `WithGRPCDialOptions(grpc.WithBlock())`, `getConnArray` might be blocked for
+	// a while when the corresponding conn array is uninitialized. However, since tidb won't set this option, we just
+	// keep `getConnArray` synchronous for now.
+	connArray, err := c.getConnArray(addr, true)
+	if err != nil {
+		cb.Invoke(nil, err)
+		return
+	}
+
+	var (
+		entry = &batchCommandsEntry{
+			ctx:           ctx,
+			req:           batchReq,
+			cb:            cb,
+			forwardedHost: req.ForwardedHost,
+			canceled:      0,
+			err:           nil,
+			pri:           req.GetResourceControlContext().GetOverridePriority(),
+			start:         time.Now(),
+		}
+		stop func() bool
+	)
+
+	// defer post actions
+	entry.cb.Inject(func(resp *tikvrpc.Response, err error) (*tikvrpc.Response, error) {
+		if stop != nil {
+			stop()
+		}
+
+		elapsed := time.Since(entry.start)
+
+		// batch client metrics
+		if sendLat := atomic.LoadInt64(&entry.sendLat); sendLat > 0 {
+			metrics.BatchRequestDurationSend.Observe(time.Duration(sendLat).Seconds())
+		}
+		if recvLat := atomic.LoadInt64(&entry.recvLat); recvLat > 0 {
+			metrics.BatchRequestDurationRecv.Observe(time.Duration(recvLat).Seconds())
+		}
+		metrics.BatchRequestDurationDone.Observe(elapsed.Seconds())
+
+		// rpc metrics
+		connArray.updateRPCMetrics(req, resp, elapsed)
+
+		// resource control
+		if stmtExec := ctx.Value(util.ExecDetailsKey); stmtExec != nil {
+			execDetails := stmtExec.(*util.ExecDetails)
+			atomic.AddInt64(&execDetails.WaitKVRespDuration, int64(elapsed))
+			execNetworkCollector := networkCollector{}
+			execNetworkCollector.onReq(req, execDetails)
+			execNetworkCollector.onResp(req, resp, execDetails)
+		}
+
+		// tracing
+		if spanRPC != nil {
+			if util.TraceExecDetailsEnabled(ctx) {
+				if si := buildSpanInfoFromResp(resp); si != nil {
+					si.addTo(spanRPC, entry.start)
+				}
+			}
+			spanRPC.Finish()
+		}
+		regionRPC.End()
+
+		// codec
+		if useCodec && err == nil {
+			resp, err = c.option.codec.DecodeResponse(req, resp)
+		}
+
+		return resp, WrapErrConn(err, connArray)
+	})
+
+	stop = context.AfterFunc(ctx, func() {
+		logutil.Logger(ctx).Debug("async send request cancelled (context done)", zap.String("to", addr), zap.Error(ctx.Err()))
+		entry.error(ctx.Err())
+	})
+
+	batchConn := connArray.batchConn
+	if val, err := util.EvalFailpoint("mockBatchCommandsChannelFullOnAsyncSend"); err == nil {
+		mockBatchCommandsChannelFullOnAsyncSend(ctx, batchConn, cb, val)
+	}
+	select {
+	case batchConn.batchCommandsCh <- entry:
```

Comment:
Are there any metrics that can help us diagnose if the channel if full?

Reply:
`tidb_tikvclient_batch_request_duration_seconds_bucket{step="send"}` can show the tail latency from entry creation to sending, but it includes the batch wait time. Currently no metric can indicate the channel full directly.

Reply:
Yeah I'm just wondering if we need another one